Add very busy expression analysis (#177)
eisop/checker-framework: dataflow/manual/content.tex

\flow{LiveSimple}{.33}{1}{Simple sequential program to illustrate live variable.  Intermediate analysis results are shown.}

\section{Example: Very Busy Expression Analysis}

> 

Author: Yes, there should a brief introduction here...

> 

Author: Ok, I removed the comparison part

\\
The main class is \code{org.checkerframework.dataflow.cfg.playground.BusyExpressionPlayground}.

\textbf{Abstract values.} The \code{BusyExprValue} is a very busy expression
 (which is represented by a node) wrapper turning node into
abstract value. The node can be any \code{BinaryOperationNode}, such as \code{NumericalAdditionNode}
 and \code{LeftShiftNode}.

\textbf{The store.} The busy expression store \code{BusyExprStore} has a field
\code{busyExprValueSet} which contains a set of \code{BusyExprValue}s.
 If a node is \code{BinaryOperationNode}, \code{addUseInExpression(Node)}
 will recursively analyze the rhs and lhs of the node to determine
 if they are also \code{BinaryOperationNode}. We then use \code{putBusyExpr(BusyExprValue)}
 to put the expression in the set. \code{killBusyExpr(Node)} is used to
 kill all expressions in set containing the variable that is re-assigned.

\textbf{The transfer function.} The transfer function \code{BusyExprTransfer}
initializes empty stores at normal and exceptional exit blocks (because this
is a backward transfer function). The transfer function visits assignment, method invocation,

>  return statement to update the information of busy expressions for each node in the stores.

Reviewer: > for each node in the stores

I think it's updating values in the stores

Author: fixed it.

\flow{LiveSimple}{.33}{1}{Simple sequential program to illustrate live variable.  Intermediate analysis results are shown.}

\section{Example: Very Busy Expression Analysis}

Unlike the live variable analysis which combines abstract states from successors,
very busy expression analysis is a backward analysis that intersects abstract states from successors.

> An expression is very busy if no matter what path is taken, the expression is used before any the variables

Reviewer: ```suggestion
An expression is very busy if no matter what path is taken, the expression is used before any of the variables
```

Author: Thanks!

Author: sure, I have already changed it. I just noticed Alex made some changes to the cfg-visualize code, so I haven't pushed the changes yet.

> An expression is very busy if no matter what path is taken, the expression is used before any the variables

Reviewer: or 
```suggestion
An expression is very busy if no matter what path is taken, the expression is used before any variables
```
Maybe this version is better.

Author: thanks!

\flow{LiveSimple}{.33}{1}{Simple sequential program to illustrate live variable.  Intermediate analysis results are shown.}

\section{Example: Very Busy Expression Analysis}

Very busy expression analysis is a backward analysis that intersects abstract states from successors.
An expression is very busy if no matter what path is taken, the expression is used before any variables
occurring in it are redefined.
\\
The main class is \code{org.checkerframework.dataflow.cfg.playground.BusyExpressionPlayground}.

\textbf{Abstract values.} The \code{BusyExprValue} is a very busy expression, represented by a node.
The node can be any \code{BinaryOperationNode}, such as \code{NumericalAdditionNode}
and \code{LeftShiftNode}.

\textbf{The store.} The busy expression store \code{BusyExprStore} has a field
\code{busyExprValueSet} which contains a set of \code{BusyExprValue}s.

> If a node is \code{BinaryOperationNode}, \code{addUseInExpression(Node)}

Reviewer 2: ```suggestion
If a node is a \code{BinaryOperationNode}, \code{addUseInExpression(Node)}
```

> will recursively analyze the rhs and lhs of the node to determine

Reviewer 2: ```suggestion
will recursively analyze the subexpressions of the node to determine
```

if they are nested \code{BinaryOperationNode}s. We then use \code{putBusyExpr(BusyExprValue)}
to put the expression in the set. \code{killBusyExpr(Node)} is used when a variable is re-assigned.

\textbf{The transfer function.} The transfer function \code{BusyExprTransfer}
initializes empty stores at normal and exceptional exit blocks (because this
is a backward transfer function). The transfer function visits assignment, method invocation,

> return statement to update the busy expression values in the stores.

Reviewer 2: ```suggestion
and return statement nodes to update the busy expression values in the stores.
```
